### PR TITLE
Log query

### PR DESCRIPTION
**Pull request by bquenin**
#11

In order to test this PR:

```
./swarm stop
./swarm pull
./swarm start --min
```

Wait 30 secs to let the logs stream in, then:

```
make test
```

You can also test the CLI:

```
make
amplifier
amp logs
amp logs --short
amp logs --n 10
amp logs --n 10 --from 0 --service_name kafka
amp logs --service_id 789456123
amp logs --message error
```

### CODE REVIEW

File: api/rpc/logs/logs.proto [hunk 1]
```diff
@@ -0,0 +1,26 @@
+syntax = "proto3";
+
+package logs;
```
Comment (bertrand-quenin) [inline api/rpc/logs/logs.proto]: I use the package 'logs' (plural) because using 'log' (singludar) conflicts with the builtin go 'log' package.

File: api/server/logs.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
+package server
```
Comment (bquenin) [inline api/server/logs.go]: I moved this file in the 'server' package to avoid a cyclic dependency issue between 'log' and 'server' packages importing each other.

**Comment by chrisccoy** [comment]
Overall LGTM. The test passed. I did notice a logs_test.go module that is part of the PR, but not part of the "make test". When I tried to run it I got this:

```
# github.com/appcelerator/amp/api/rpc/logs
import cycle not allowed in test
package github.com/appcelerator/amp/api/rpc/logs (test)
    imports github.com/appcelerator/amp/api/server
    imports github.com/appcelerator/amp/api/rpc/logs

FAIL    github.com/appcelerator/amp/api/rpc/logs [setup failed]
```

Should it be there?

**Comment by subfuzion** [comment]
@bertrand-quenin PRs should have a reference to an issue and some kind of description that summarizes what to expect out of the PR. Here are examples: https://github.com/appcelerator/amp/pull/7, https://github.com/appcelerator/amp/pull/17, https://github.com/appcelerator/amp/pull/6. Thanks.

File: api/rpc/logs/logs.proto (hunk 1 above)
Comment (subfuzion) [inline api/rpc/logs/logs.proto]: Ah, right, makes sense.

File: api/server/logs.go (hunk 2 above)
Comment (subfuzion) [inline api/server/logs.go]: Why does this need to access to anything in `server`? Is it because of `ES`? That can be passed as a parameter, right?

File: data/elasticsearch/elasticsearch.go [hunk 3]
```diff
@@ -23,26 +24,67 @@ func (es *ElasticSearch) Connect(url string) error {
 	return err
 }
 
-// Index stores a document inside elastic search
-func (es *ElasticSearch) Index(esIndex string, esType string, body interface{}) error {
+// CreateIndexIfNotExists Creates an index if it doesn't already exists
```
Comment (subfuzion) [inline data/elasticsearch/elasticsearch.go:27]: We are going to use this package for fetching data only (to the best of my knowledge). The log worker should be responsible for ensuring the index exists for the data it will be writing. Are you aware of any other use cases for elasticsearch?

File: data/elasticsearch/elasticsearch.go [hunk 4]
```diff
@@ -23,26 +24,67 @@ func (es *ElasticSearch) Connect(url string) error {
...
-	return err
+	if err != nil {
+		// TODO: Handle error
+		panic(err)
+	}
 }
 
-// All returns all the documents for a given index
-func (es *ElasticSearch) All(esIndex string) ([]*elastic.SearchHit, error) {
+// All Returns all the documents for a given index
```
Comment (subfuzion) [inline data/elasticsearch/elasticsearch.go]: It looks better the way it was ("returns").

File: data/elasticsearch/elasticsearch.go [hunk 5]
```diff
@@ -23,26 +24,67 @@ func (es *ElasticSearch) Connect(url string) error {
...
 	if err != nil {
-		return nil, err
+		// TODO: Handle error
+		panic(err)
 	}
-	return searchResult.Hits.Hits, nil
+	return searchResult.Hits.Hits
+}
+
+// GetNative returns the native elastic search client
```
Comment (subfuzion) [inline data/elasticsearch/elasticsearch.go]: Update comment (GetNative -> GetClient).

File: data/elasticsearch/elasticsearch.go (hunk 3 above)
Comment (bquenin) [inline data/elasticsearch/elasticsearch.go:27]: You're right. This is just from previous experimentation, but this code is not called. Do you want me to remove it ?

File: api/server/logs.go (hunk 2 above)
Comment (bquenin) [inline api/server/logs.go]: Yes because of ES. Passed as a parameter to what ?

File: Makefile [hunk 6]
```diff
@@ -44,11 +44,12 @@ clean:
 	@rm -f $$(which amp)
 
 test:
-	@go test -v $(REPO)/data/storage/etcd
-	@go test -v $(REPO)/data/influx
-	@go test -v $(REPO)/api/rpc/project
-	@go test -v $(REPO)/api/rpc/service
-	@go test -v $(REPO)/api/rpc/stat
+#	@go test -v $(REPO)/data/storage/etcd
```
Comment (subfuzion) [inline Makefile]: Don't want to merge a PR with tests commented out. It's better to tell the PR reviewer in your description to just cd to `api/rpc/logs` and run `go test`.

**Comment by freignat91** [comment]
LGTM

File: data/elasticsearch/elasticsearch.go (hunk 3 above)
Comment (subfuzion) [inline data/elasticsearch/elasticsearch.go:27]: I had removed it in the previous PR that was merged and you added it back. I guess you can remove it in your next PR.

File: cli/app.go [hunk 7]
```diff
@@ -53,3 +64,54 @@ func (a *AMP) Status() {
 		fmt.Println("Status")
 	}
 }
+
+// Logs fetches the logs
+func (a *AMP) Logs(cmd *cobra.Command) error {
```
Comment (subfuzion) [inline cli/app.go:69]: This isn't how I intented `app.go` go get used. It is the job of `cmd/amp/main.go`'s job to parse the command line then call `a.Logs(args...)` not `a.Logs(*cobra.command)`. I didn't intend for the app object to actually have to know about cobra at all. This make the app object more testable. Also, I figure if the implementation of a command takes more than a few lines, there probably should be a separate file for it (ex: `cli/log.go`). I suppose to make the intention clearer, it might be better to rename the package `cli` -> `client`. You can do that if you want as part of this (or your next PR), or I will get to it with mine.

Comment (bquenin) [inline cli/app.go:69]: Sure. Can it be done in another pull request though ? I'd like to avoid this PR to linger for a week :)

**Comment by subfuzion** [comment]
@bertrand-quenin Looking pretty good. I commented above about `app.go`. I'd like it fixed, but if you want to tackle that in the next PR, that's fine (this PR is not intended to close #11 yet, correct)?

File: cli/app.go (hunk 7 above)
Comment (subfuzion) [inline cli/app.go:69]: Yeah, I'm totally fine with that.

**Comment by subfuzion** [comment]
@bertrand-quenin Looking pretty good. I commented above about `app.go`, but you can tackle that with the next PR for this feature. 👍

**Comment by bquenin** [comment]
Thanks, you're right it doesn't close #11 yet
I need to address your comments and also add more unit tests.
I'll submit another PR when I'm back.
Thanks!
